### PR TITLE
rbac: update database permissions on startup

### PR DESCRIPTION
**Pull request by BolajiOlajide**
Closes [#45455](https://github.com/sourcegraph/sourcegraph/issues/45455)

## Test plan

Restarting the `frontend` service should seed permissions from the `internal/rbac/schema.yaml` file into the database. Removing or adding any permission to the file should automatically propagate into the database on restart.

### CODE REVIEW

File: cmd/frontend/internal/bg/update_permissions.go [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+import (
+	"context"
+
+	"github.com/sourcegraph/log"
+
+	"github.com/sourcegraph/sourcegraph/internal/database"
+	"github.com/sourcegraph/sourcegraph/internal/rbac"
+)
+
+func UpdatePermissions(ctx context.Context, logger log.Logger, db database.DB) {
```
Comment (Piszmog) [inline cmd/frontend/internal/bg/update_permissions.go:12]: Docs please

File: internal/database/permissions.go [hunk 2]
```diff
@@ -170,6 +180,39 @@ func (p *permissionStore) Delete(ctx context.Context, opts DeletePermissionOpts)
 	return nil
 }
 
+func (p *permissionStore) BulkDelete(ctx context.Context, opts []DeletePermissionOpts) error {
+	if len(opts) == 0 {
+		return nil
+	}
```
Comment (Piszmog) [inline internal/database/permissions.go:186]: Just a thought I had when looking at this - I could see it both ways of either doing nothing or returning an error.

What was your thought on going the `nil` route vs an error?

File: internal/database/permissions_test.go [hunk 3]
```diff
@@ -148,3 +148,44 @@ func TestPermissionBulkCreate(t *testing.T) {
 	assert.NotNil(t, ps)
 	assert.Len(t, ps, 5)
 }
+
+func TestPermissionBulkDelete(t *testing.T) {
```
Comment (Piszmog) [inline internal/database/permissions_test.go:152]: Would like to see the additional tests,
* No options provided as an argument
* No rows affected
* And the error cases (if possible with our framework)

File: internal/rbac/permissions.go [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+
+	var parsedSchema Schema
+	if err := yaml.Unmarshal(contents, &parsedSchema); err != nil {
+		panic(fmt.Sprintf("malformed rbac schema definition: %s", err.Error()))
+	}
+
+	return parsedSchema
+}()
+
+// comparePermissions takes two slices of permissions (one from the database and another from the schema file)
```
Comment (Piszmog) [inline internal/rbac/permissions.go:30]:
```suggestion
// ComparePermissions takes two slices of permissions (one from the database and another from the schema file)
```

File: internal/rbac/types.go [hunk 5]
```diff
@@ -0,0 +1,11 @@
+package rbac
+
+type Schema struct {
```
Comment (Piszmog) [inline internal/rbac/types.go:3]: Docs

**Review by Piszmog — APPROVED** [review]

File: internal/database/permissions.go (hunk 2 above)
Comment (BolajiOlajide) [reply internal/database/permissions.go:186]: Reverting to return an error so it's consistent with other methods though.

File: internal/database/permissions_test.go (hunk 3 above)
Comment (BolajiOlajide) [reply internal/database/permissions_test.go:152]: Added

File: internal/rbac/types.go (hunk 5 above)
Comment (BolajiOlajide) [reply internal/rbac/types.go:5]: Added

File: cmd/frontend/internal/bg/update_permissions.go (hunk 1 above)
Comment (BolajiOlajide) [reply cmd/frontend/internal/bg/update_permissions.go:17]: Added.

Comment (BolajiOlajide) [reply cmd/frontend/internal/bg/update_permissions.go:17]: Added

File: internal/rbac/types.go [hunk 6]
```diff
@@ -0,0 +1,14 @@
...
+// the RBAC system.
+type Schema struct {
+	Namespaces []Namespace `json:"namespaces"`
+}
+
+// Namespace represents a feature to be guarded by RBAC. (example: Batch Changes, Code Insights e.t.c)
+type Namespace struct {
+	Name         string   `json:"name"`
+	Actions      []string `json:"actions"`
+	DefaultApply bool     `json:"defaultApply"`
```
Comment (courier-new) [inline internal/rbac/types.go:13]: What does `DefaultApply` refer to?

**Pull request merged**